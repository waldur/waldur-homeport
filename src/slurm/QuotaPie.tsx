// This directive is based on code from https://hackernoon.com/a-simple-pie-chart-in-svg-dbdd653b6936
import * as React from 'react';

const getCoordinatesForPercent = percent => {
  const x = Math.cos(2 * Math.PI * percent);
  const y = Math.sin(2 * Math.PI * percent);
  return [x, y];
};

const getSlices = value => [
  {
    percent: value,
    color: 'Coral',
  },
  {
    percent: 1 - value,
    color: '#00ab6b',
  },
];

export const QuotaPie = ({ value }) => {
  const slices = getSlices(value);
  let cumulativePercent = 0;

  return (
    <svg viewBox="-1 -1 2.1 2.1" rotate="-90deg" height="15px">
      {slices.map((slice, index) => {
        // destructuring assignment sets the two variables at once
        const [startX, startY] = getCoordinatesForPercent(cumulativePercent);

        // each slice starts where the last slice ended, so keep a cumulative percent
        cumulativePercent += slice.percent;

        const [endX, endY] = getCoordinatesForPercent(cumulativePercent);

        // if the slice is more than 50%, take the large arc (the long way around)
        const largeArcFlag = slice.percent > 0.5 ? 1 : 0;

        // create an array and join it just for code readability
        const pathData = [
          `M ${startX} ${startY}`, // Move
          `A 1 1 0 ${largeArcFlag} 1 ${endX} ${endY}`, // Arc
          `L 0 0`, // Line
        ].join(' ');

        return <path key={index} d={pathData} fill={slice.color} />;
      })}
    </svg>
  );
};
