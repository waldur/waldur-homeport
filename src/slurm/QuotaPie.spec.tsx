import { render } from 'enzyme';
import * as React from 'react';

import { QuotaPie } from './QuotaPie';

describe('Quota pie', () => {
  it('renders SVG chart', () => {
    expect(render(<QuotaPie value={0.33} />).html()).toMatchSnapshot();
  });
});
