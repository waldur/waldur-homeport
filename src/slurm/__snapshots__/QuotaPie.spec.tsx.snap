// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quota pie renders SVG chart 1`] = `"<path d=\\"M 1 0 A 1 1 0 0 1 -0.48175367410171543 0.8763066800438635 L 0 0\\" fill=\\"Coral\\"></path><path d=\\"M -0.48175367410171543 0.8763066800438635 A 1 1 0 1 1 1 -2.4492935982947064e-16 L 0 0\\" fill=\\"#00ab6b\\"></path>"`;
